k8s-operator/sessionrecording/ws/conn.go

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+// package ws has functionality to parse 'kubectl exec' sessions streamed using
+// WebSockets protocol.
+package ws
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/remotecommand"
+	srconn "tailscale.com/k8s-operator/sessionrecording/conn"
+	"tailscale.com/k8s-operator/sessionrecording/tsrecorder"
+	"tailscale.com/sessionrecording"
+	"tailscale.com/util/multierr"
+)
+
+// New returns a wrapper around net.Conn that intercepts reads and writes for a
+// websocket streaming session over the provided net.Conn, parses the data as
+// websocket messages and sends message payloads for STDIN/STDOUT streams to a
+// tsrecorder instance using the provided client. Caller must ensure that the
+// session is streamed using WebSockets protocol.
+func New(c net.Conn, rec *tsrecorder.Client, ch sessionrecording.CastHeader, log *zap.SugaredLogger) srconn.Conn {
```

Comment:
IIUC the `net.Conn` we receive here is between the `kubectl` client and the operator. The `http.ReverseProxy` handles the `net.Conn` between the operator and upstream (API server). So when we're reading from this conn we're reading from the client, and writes are writing to the client. It's already a fairly detailed doc comment (thanks!), but it might be helpful to spell that out in a bit more detail because piecing together the full picture requires traversing quite a bit of the call stack.

With that in mind, I'm noticing that we never call the recorder in the `Read` function, which means we don't capture stdin (which I recall is intentional to avoid capturing sensitive info like passwords etc). Does that mean we also don't capture the commands for interactive sessions, only their output? (Sorry I probably forgot the answer from when you implemented SPDY)

Reply:
As discussed in a separate thread - we do capture commands output to terminal, but that can be bypassed. Currently stdin is not captured to avoid capturing hidden input like passwords - we will ask for user feedback to understand if this should be changed or made configurable